### ISSUE TITLE
Simulator vs. device performance

### ISSUE THREAD

**Issue by HarshilShah** [maxkonovalov/MKRingProgressView#25]
I'm trying to use the example application with the Xcode 9 and iOS 11 GMs here, and seeing significant performance differences between the two.

On my mid-2014 15-inch MacBook Pro, the example app absolutely flies, working perfectly, pretty much. Meanwhile, on my 6s Plus, it's dropping frames left, right, and centre. I tried messing with the usual suspects (clear memory, low power mode, etc.) but device just keeps dropping frames.

I know there's a performance difference between simulator and device, but it usually tends to be in the device's favour, especially with graphics stuff. There's also an OpenGL simulator bug in the GM, which might be at play here: https://twitter.com/xenadu02/status/911463433521860609

Any ideas for increasing on device performance? I tried but can't seem to find any way to toggle CPU-based rendering like with CIContext, for example.

Thanks

**Comment by maxkonovalov** [comment]
Hey @HarshilShah, thanks for pointing this out! There used to be a performance issue with generating the gradient image, but I'm not sure if it could affect the frame rate so much. Can you check if you have the same issue on a **Release** config?
More info here - https://github.com/maxkonovalov/MKGradientView#performance

**Comment by maxkonovalov** [comment]
Yup, looks like a CPU-bound issue then, as everything is drawn using CGContext. Definitely need to profile and optimize that. Don't know when I can get my hands on it though, so if you have any ideas about fixing it, any help would be appreciated!
Btw I tested the demo on a iPhone 7+ device with iOS 11 and the performance was not bad at all.

**Comment by HarshilShah** [comment]
Yup, I’ve got a couple of ideas I’ll be trying. Also I do need to upgrade my phone 😅

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by HarshilShah** [comment]
Btw, here’s what it looks like on my end. Kinda surprised iOS screen recording works even when there isn’t even enough perf to render frames in time, but this is exactly what I’m seeing: 
[ScreenRecording_09-25-2017 15:04.MP4.zip](https://github.com/maxkonovalov/MKRingProgressView/files/1328996/ScreenRecording_09-25-2017.15.04.MP4.zip)

(Zipped because GitHub doesn't support .mp4)

**Comment by maxkonovalov** [comment]
Thanks for posting the video! I thought the issue was much worse actually))

**Comment by HarshilShah** [comment]
It gets worse with progress, basically. More than one rotation of any ring makes it way worse

**Comment by maxkonovalov** [comment]
Hi @HarshilShah, I updated the lib with some optimizations - can you please check v2.0 and let me know if the performance improved for you.
